partiql-parser/src/preprocessor.rs

```diff
@@ -0,0 +1,680 @@
+use partiql_source_map::location::ByteOffset;
+use regex::{Regex, RegexSet, RegexSetBuilder};
+
+use std::collections::VecDeque;
+
+use std::ops::Range;
+
+use crate::result::LexError;
+
+use crate::lexer::{InternalLexResult, LexResult, PartiqlLexer, Spanned, Token};
+
+use crate::token_parser::{BufferedToken, TokenParser};
+use partiql_source_map::line_offset_tracker::LineOffsetTracker;
+
+/// A single "function expression" argument match.
+#[derive(Debug, Clone)]
+pub(crate) enum FnExprArgMatch<'a> {
+    /// Any 1 [`Token`] that is not function punctuation (i.e., '(', ')', ',') and not a keyword
+    Expr,
+    /// 0 or more [`Token`]s that are not function punctuation (i.e., '(', ')', ',') and not a keyword
+    Any,
+    /// Explicitly match a single [`Token`]
+    Match(Token<'a>),
+    /// Explicitly match a [`Token`] that is a keyword that represents a 'named' argument.
+    ///
+    /// For example, the `for` in `substring(x for 2)`
+    NamedArgKw(Token<'a>),
+    /// Match a regex against an identifier that represents a 'named' argument.
+    ///
+    /// For example, the `leading` in `trim(leading ' ' from x)`
+    NamedArgId(Regex),
+    /// Synthesize a value for a named argument if one is not explicitly provided. Used for default values.
+    ///
+    /// For example, in `trim(leading from x)`, there is a synthesized `' '` that is the target of the
+    /// `leading` such that it is interpreted as `trim(leading ' ' from x)`
+    Synthesize(Token<'a>),
+}
+
+/// A "function expression" argument match list.
+pub(crate) type FnExprArgList<'a> = Vec<FnExprArgMatch<'a>>;
+
+/// A "function expression" match.
+#[derive(Debug, Clone)]
+pub(crate) struct FnExpr<'a> {
+    /// Name(s) of "function expression"
+    pub fn_names: Vec<&'a str>,
+    /// A collection of possible argument patterns for this function expression.
+    pub patterns: Vec<FnExprArgList<'a>>,
+}
+
+mod built_ins {
+    use super::*;
+    use regex::Regex;
+
+    use FnExprArgMatch::{
+        Any as EAny, Expr, Match as T, NamedArgId as Id, NamedArgKw as Kw, Synthesize as Syn,
+    };
+
+    const TRIM_SPECIFIER: &str = "(?i:leading)|(?i:trailing)|(?i:both)";
+
+    pub(crate) fn built_in_trim() -> FnExpr<'static> {
+        let re = Regex::new(TRIM_SPECIFIER).unwrap();
+        FnExpr {
+            fn_names: vec!["trim"],
+            #[rustfmt::skip]
+            patterns: vec![
+                // e.g., trim(leading 'tt' from x) => trim("leading": 'tt', "from": x)
+                vec![Id(re.clone()), Expr, EAny, Kw(Token::From), Expr, EAny],
```

Comment:
`..., Expr, EAny, ...` is essentially like a regex `(Expr)+`. That is, it requires at least one token (that's the `Expr` part) and potentially more (the `EAny`) part until there is a further match in the pattern. This is because this parser parses an "island grammar"; In other words, it doesn't actually know the full expression grammar, it just knows some landmarks and passes all other tokens.

I tried condensing them into one state, but it is difficult with the statelessness of the parsing algorithm, and required keeping a separate flag or or count to pass between iterations.

I also tried different names for them, e.g., RequiredExpr and ZeroOrMoreExpr, etc. But I didn't find them more descriptive; let me know if you do. Maybe `AnyOne`, `AnyZeroMore` ?

Reply:
Thanks for explaining, I think the names are fine, maybe we can just add a short comment regarding your explanation?

Reply:
I think the explanation here makes sense. Adding the explanation you provided here regarding `Expr` combined with `EAny` as a comment would be helpful. Also think the original names are clear + descriptive enough.

I'm curious what would happen if the two arguments were swapped (i.e. `..., EAny, Expr,...`)? Would that represent some other regex? Or would it represent the same regex `(Expr)+`)?

Reply:
It would not be the same. It would probably be more like the regex `(Expr)?(Expr)`